### ISSUE TITLE
Can shapely 2.0 support unpickling of shapely 1 object?

### ISSUE THREAD

**Issue by zhengwy888** [shapely/shapely#1613]
## Expected behavior and actual behavior.
shapely 2.0 can't unpickle objects generated from shapely 1.8. I am guessing this is intentional because shapely 2.0 objects are immutable. But we have many existing pickles that's also difficult to convert to new shapely 2.0 geometry class. Wondering if the library can provide a backward compatible method to load old pickles? 

## Steps to reproduce the problem.

in a shapely 1.8 environment:
```py
import shapely.geometry
import pickle
shape = shapely.geometry.Polygon([[0,0],[3,3],[2,2]])
bytes = pickle.dumps(shape)
```

in a shapely 2.0 enviroment
```py
import shapely
import pickle
pickle.loads(<the bytes from above>)
```
And it will throw error `_pickle.UnpicklingError: state is not a dictionary`, because `__setstate__` is not implemented. 

## Operating system
Ubuntu 20

## Shapely version and provenance
2.0b2 installed from PyPI using pip

**Comment by jorisvandenbossche** [comment]
I think the simple explanation is that we didn't try to make unpickling backwards compatible, and indeed because the simple solution is not possible, as you mention: geometry objects are now immutable, and so the problem is that if we add a `__setstate__` to handle the state of 1.x shapely objects, at that moment the geometry object already has been created (as an empty geometry, by calling the class constructor without arguments), and so we can't update this object inplace with the WKB that is stored in the state of 1.x pickles. 

The pickle module provides quite some tools to tweak this process, or to provide custom overrides or Unpickler classes (https://docs.python.org/3/library/pickle.html), so it might be that there is something possible to provide a function that can load those older pickles. But I am not super familiar with this. Experimenting with this would certainly be welcome!

---

The pickle files are under the hood just using WKB as the serialization format. So one workaround is to load your pickle files in Shapely 1.8, store them as pure wkb instead, and then load that in Shapely 2.0. 

Something like:

```python
import pickle
import shapely.wkb

# creating a test pickle file
from shapely.geometry import Point
p = Point(1, 1)
with open("point1.pickle", "wb") as f:
    pickle.dump(p, f)

# convert all pickle files to pure wkb files (could be done in a loop)
with open("point1.pickle", "rb") as f:
    geom = pickle.load(f)
with open("point1.wkb", "wb") as f:
    f.write(shapely.wkb.dumps(geom))
```

And then with shapely 2.0, we can read that:

```python
import shapely
import pickle

# unpickling fails as expected
>>> with open("point1.pickle", "rb") as f:
...     geom = pickle.load(f)
..
UnpicklingError: state is not a dictionary

# but we can read the pure WKB file
>>> with open("point1.wkb", "rb") as f:
...     geom = shapely.from_wkb(f.read())
>>> geom
<POINT (1 1)>

**Comment by zhengwy888** [comment]
we have many different pickles that are stored in nested data structures, so simply converting them to binary wkb is non-trivial either. 

I got around my issue without upgrading the shapely library, but for anyone who wants to investigate this in the future, I would say overwriting the [_Unpickler](https://github.com/python/cpython/blob/main/Lib/pickle.py) seemed to be a promising route, but at the expense of some performance. Inside `load_build`, pop the stack, create a new Shapely 2.0 object, and then append it back to the stack. 

the normal Unpickler only provides `find_class` method for overwrites, which allows you to define a custom class to load the wkb, but I got stuck in how to make the custom class behaves exactly like the new Shapely 2.0 classes. 
feel free to close this issue.
